homeassistant/components/light/fibaro.py

```diff
@@ -0,0 +1,165 @@
+"""
+Support for Fibaro lights.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/light.fibaro/
+"""
+
+import logging
+import threading
+
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS, ATTR_HS_COLOR, ATTR_WHITE_VALUE, ENTITY_ID_FORMAT,
+    SUPPORT_BRIGHTNESS, SUPPORT_COLOR, SUPPORT_WHITE_VALUE, Light)
+import homeassistant.util.color as color_util
+from homeassistant.components.fibaro import (
+    FIBARO_CONTROLLER, FIBARO_DEVICES, FibaroDevice)
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['fibaro']
+
+
+def scaleto255(value):
+    """Scale the input value from 0-100 to 0-255."""
+    # Fibaro has a funny way of storing brightness either 0-100 or 0-99
+    # depending on device type (e.g. dimmer vs led)
+    if value > 98:
+        value = 100
+    return max(0, min(255, ((value * 256.0) / 100.0)))
+
+
+def scaleto100(value):
+    """Scale the input value from 0-255 to 0-100."""
+    # Make sure a low but non-zero value is not rounded down to zero
+    if 0 < value < 3:
+        return 1
+    return max(0, min(100, ((value * 100.4) / 255.0)))
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Perform the setup for Fibaro controller devices."""
+    if discovery_info is None:
+        return
+
+    add_entities(
+        [FibaroLight(device, hass.data[FIBARO_CONTROLLER])
+         for device in hass.data[FIBARO_DEVICES]['light']], True)
+
+
+class FibaroLight(FibaroDevice, Light):
+    """Representation of a Fibaro Light, including dimmable."""
+
+    def __init__(self, fibaro_device, controller):
+        """Initialize the light."""
+        self._supported_flags = 0
+        self._last_brightness = 0
+        self._color = (0, 0)
+        self._brightness = None
+        self._white = 0
+
+        self._update_lock = threading.RLock()
+        if 'levelChange' in fibaro_device.interfaces:
+            self._supported_flags |= SUPPORT_BRIGHTNESS
+        if 'color' in fibaro_device.properties:
+            self._supported_flags |= SUPPORT_COLOR
+        if 'setW' in fibaro_device.actions:
+            self._supported_flags |= SUPPORT_WHITE_VALUE
+        super().__init__(fibaro_device, controller)
+        self.entity_id = ENTITY_ID_FORMAT.format(self.ha_id)
+
+    @property
+    def brightness(self):
+        """Return the brightness of the light."""
+        return scaleto255(self._brightness)
+
+    @property
+    def hs_color(self):
+        """Return the color of the light."""
+        return self._color
+
+    @property
+    def white_value(self):
+        """Return the white value of this light between 0..255."""
+        return self._white
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return self._supported_flags
+
+    def turn_on(self, **kwargs):
+        """Turn the light on."""
+        with self._update_lock:
```

Comment:
This could block threads in the thread pool. That's not allowed. A non blocking acquire could work. What are we trying to protect?

Reply:
The issue I'm trying to solve is that while we're still executing the turn_on command, another thread is receiving notifications from the fibaro hub and processing them, e.g. there can be calls to the update function. Which can mess up the data a bit and cause some havoc. This is especially troubling in the case of a turned off colored light, which has to issue two consecutive rest api calls to the fibaro hub, both of them resulting one or multiple status updates from the fibaro hub. Since neither function has infinite loops and only of them has actual IO, I don't see how it could lock up a thread permanently, but I'm very open to other ideas of solving this race condition.

Reply:
The race is real, I had the same issue with LIFX. I am not sure how to best solve it for a threaded platform. As is, I guess it could get ugly if the hub gets unavailable.

Reply:
So this should become `async_turn_on` and Fibaro needs to maintain an async lock. We can't solve this on the Home Assistant level as we don't know exactly how the resources need to be treated that need locking (is it per device, per hub etc).